lib/pmdtester/builders/pmd_report_builder.rb

```diff
@@ -0,0 +1,94 @@
+require 'fileutils'
+require '../cmd'
+include PmdTester
+module PmdTester
+  class PmdReportBuilder
+    def initialize(branch_config, projects, local_git_repo, pmd_branch_name, rule_sets)
+      @branch_config = branch_config
+      @projects = projects
+      @local_git_repo = local_git_repo
+      @pmd_branch_name = pmd_branch_name
+      @rule_sets = rule_sets
+      @pwd = Dir.getwd
+    end
+
+    def create_repositories_dir
+      @repositories_dir = "#@pwd/target/repositories"
+      FileUtils.mkdir(@repositories_dir) unless File.directory?(@repositories_dir)
+    end
+
+    def execute_reset_cmd(type, tag)
+      case type
+      when "git"
+        reset_cmd = "git reset --hard #{tag}"
+      when "hg"
+        reset_cmd = "hg up #{tag}"
+      else
+        raise Exception, "Unknown #{type} repository"
+      end
+
+      Cmd.execute(reset_cmd)
+    end
+
+    def get_projects
+      puts 'Cloning projects started'
+
+      create_repositories_dir
+
+      @projects.each do |project|
+        path = @repositories_dir + project.name
+        clone_cmd = "#{project.type} clone #{project.connection} #{path}"
```

Comment:
this doesn't support having the repository already cloned (ie: having it cached under Travis), which I forsee is going to happen... I'd at least have a TODO to remind ourselves later

Reply:
It currently only clones, if the repository directory (`path`) doesn't exist yet - see the following lines on master (https://github.com/pmd/pmd-regression-tester/blob/master/lib/pmdtester/builders/pmd_report_builder.rb#L40-L43). However, this will probably see more tweaking in the future...